bytecomp/switch.ml

```diff
@@ -236,82 +249,132 @@ let pret chan = function
+                low_loc = _;
+                low = _;
+                high_plus_one_loc = h_plus_one_loc;
+                high = h;
+                action_index = _;
+              } = r.(len1-2)
+              in
               if h+1 < l1 then
-                h+1
+                h_plus_one_loc, h+1
               else
-                l1
+                l1_loc, l1
             end else
-              l1
-          and h =
+              l1_loc, l1
+          and h_plus_one_loc, h =
             if 1 < len2-1 then begin
-              let l,_,_ = c2.(1) in
+              let {
+                low_loc = l_loc;
+                low = l;
+                high_plus_one_loc = _;
+                high = _;
+                action_index = _;
+              } = c2.(1) in
               if h2+1 < l then
-                l-1
+                l_loc, l-1
               else
-                h2
+                h2_plus_one_loc, h2
             end else
-              h2 in
-          r.(len1-1) <- (l,h,act1) ;
+              h2_plus_one_loc, h2
+          in
+          r.(len1-1) <- {
+            low_loc = l_loc;
+            low = l;
+            high_plus_one_loc = h_plus_one_loc;
+            high = h;
+            action_index = act1;
+          };
           for i=1 to len2-1  do
             r.(len1-1+i) <- c2.(i)
           done ;
           r
-        else if h1 > l1 then
+        end else if h1 > l1 then begin
           let r = Array.make (len1+len2) c1.(0) in
           for i = 0 to len1-2 do
             r.(i) <- c1.(i)
           done ;
-          r.(len1-1) <- (l1,l2-1,act1) ;
+          r.(len1-1) <- {
+            low_loc = l1_loc;
+            low = l1;
+            high_plus_one_loc = l2_loc;
+            high = l2 - 1;
+            action_index = act1;
+          };
           for i=0 to len2-1  do
             r.(len1+i) <- c2.(i)
           done ;
           r
-        else if h2 > l2 then
+        end else if h2 > l2 then begin
           let r = Array.make (len1+len2) c1.(0) in
           for i = 0 to len1-1 do
             r.(i) <- c1.(i)
           done ;
-          r.(len1) <- (h1+1,h2,act2) ;
+          r.(len1) <- {
+            low_loc = h1_plus_one_loc;
+            low = h1 + 1;
+            high_plus_one_loc = h2_plus_one_loc;
+            high = h2;
+            action_index = act2;
+          };
           for i=1 to len2-1  do
             r.(len1+i) <- c2.(i)
           done ;
           r
-        else
+        end else begin
           Array.append c1 c2
+        end
 
 
-  let coupe_inter i j cases =
+  let slice_cases i j (cases : cases) =
     let lcases = Array.length cases in
-    let low,_,_ = cases.(i)
-    and _,high,_ = cases.(j) in
-    low,high,
-    Array.sub cases i (j-i+1),
-    case_append (Array.sub cases 0 i) (Array.sub cases (j+1) (lcases-(j+1)))
+    let { low; _ } = cases.(i)
+    and { high; _ } = cases.(j) in
+    assert (j >= i);
+    let pivot_case_for_loc = cases.(i + ((j - i) / 2)) in
```

Comment:
Why use `i + (j - i) / 2` here, is this just an arbitrary choice?

Reply:
Some choice has to be made, and I think this seemed reasonable as a range bisection was occurring during the search (as far as I could tell).